v2/internal/testcommon/kube_per_test_context.go

```diff
@@ -456,6 +456,42 @@ func (tc *KubePerTestContext) DeleteResourcesAndWait(objs ...client.Object) {
 	}
 }
 
+// DeleteResourcesAndWait deletes the resources in K8s and waits for them to be deleted
+func (tc *KubePerTestContext) DeleteResourcesAndWait(objs ...client.Object) {
+	// Protect against deleting resources that have a parent-child relationship in the same request. This is perfectly
+	// fine in the real world, but in many of our recording envtests we can get into a situation where there's a race
+	// during deletion that causes HTTP replay issues. The sequence of events is:
+	// 1. Delete parent resource and child resource at the same time.
+	// 2. During recording, child deletion never sees a successful (finished) DELETE request because parent is deleted so soon
+	//    after the child that we just record a successful DELETE for the parent and don't bother sending the final child
+	//    DELETE that would return success.
+	// 3. During replay, the race is how quickly the parent deletes and how many requests the child has a chance to send
+	//    in that time. If the parent deletes slowly the child might try to send more requests than we actually have recorded
+	//    (because none of them represent a terminal "actually deleted" state), which will cause a test failure.
+	oids := make(map[types.UID]client.Object)
+	for _, obj := range objs {
+		oids[obj.GetUID()] = obj
+	}
+
+	// Note: This is imperfect. If you have a situation like A owns B owns C and you pass just A and C here, we won't detect it
+	var roots []client.Object
+
+	for _, obj := range objs {
+		refs := obj.GetOwnerReferences()
+		foundOwner := false
+		for _, ref := range refs {
+			if _, ok := oids[ref.UID]; ok {
+				foundOwner = true
```

Comment:
maybe `break` after when owner is found? As we don't want to keep iterating?

Reply:
Done